File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left=0;
        int maxLen=0;
        int n=s.length();

        HashSet<Character> seen = new HashSet<>();
        
        
        for(int right=0;right<n;right++){
            while(seen.contains(s.charAt(right))){
                seen.remove(s.charAt(left));
                left++;
                
            }
            seen.add(s.charAt(right));
            int len=right-left+1;
            maxLen=Math.max(maxLen, len);

            
            
            
            
        }
        return maxLen;
        
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        left=0
        maxfreq=0
        count={}
        longest=0

        for right in range (len(s)):

            count[s[right]]=count.get(s[right],0)+1
            maxfreq=max(maxfreq, count[s[right]])

            while (right-left+1)-maxfreq>k:
                count[s[left]]-=1
                left+=1

            longest=max(longest, right-left+1)
        return longest
                





        